### PR TITLE
Handle hvdcs in rao

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
Only PstRangeActions are optimized

**What is the new behavior (if this is a feature change)?**
Hvdcs are also optimized.

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
No

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java [hunk 1]
```diff
@@ -229,7 +228,7 @@ private boolean exceedMaxNumberOfRaPerTso(NetworkActionCombination naCombination
 
         // get max number of network action which can still be activated, per Tso
         tsos.forEach(tso -> {
-            int activatedPstsForTso = (int) fromLeaf.getRangeActions().stream().filter(rangeAction -> rangeAction instanceof PstRangeAction && hasRangeActionChangedComparedToPrePerimeter(fromLeaf, rangeAction)).count();
+            int activatedPstsForTso = (int) fromLeaf.getRangeActions().stream().filter(rangeAction -> hasRangeActionChangedComparedToPrePerimeter(fromLeaf, rangeAction)).count();
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java:231]: rename this to activatedRangeActionsForTso

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/linear_optimisation/BestTapFinder.java [hunk 2]
```diff
@@ -53,15 +53,20 @@ public static RangeActionResult find(RangeActionResult rangeActionResult,
         Map<PstRangeAction, Integer> bestTaps = new HashMap<>();
         Map<PstRangeAction, Map<Integer, Double>> minMarginPerTap = new HashMap<>();
 
-        Set<PstRangeAction> pstRangeActions = rangeActionResult.getOptimizedTaps().keySet();
+        Set<PstRangeAction> pstRangeActions = new HashSet<>();
+        optimizedSetPoints.keySet().forEach(ra -> {
+            if (ra instanceof PstRangeAction) {
+                pstRangeActions.add((PstRangeAction) ra);
+            }
+        });
```
Comment (pet-mit) [inline ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/linear_optimisation/BestTapFinder.java:61]: use keySet.stream.filter.collect instead

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java [hunk 3]
```diff
@@ -92,6 +92,10 @@ void setAvailableRangeActions(Set<RangeAction> rangeActions) {
         availableRangeActions = rangeActions;
     }
 
+    void setPrePerimeterRangeActionSetPoints(Map<RangeAction, Double> prePerimeterRangeActionSetPoints) {
+        this.prePerimeterRangeActionSetPoints = prePerimeterRangeActionSetPoints;
+    }
+
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:98]: I'm not a fan of exposing this method to the whole package (even though you only need it for tests).
can you use mocks in the unit tests instead ?

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/linear_optimisation/IteratingLinearOptimizer.java [hunk 4]
```diff
@@ -147,8 +159,14 @@ private IteratingLinearOptimizerResult createResult(FlowResult flowResult,
     }
 
     private RangeActionResult roundResult(RangeActionResult rangeActionResult, Network network, IteratingLinearOptimizerResult previousResult) {
+        Map<RangeAction, Double> roundedSetPoints = new HashMap<>(rangeActionResult.getOptimizedSetPoints());
+        // Round set points to the closest integer for non-pst range actions
+        for (RangeAction rangeAction : rangeActionResult.getOptimizedSetPoints().keySet().stream().filter(ra -> !(ra instanceof PstRangeAction)).collect(Collectors.toSet())) {
+            roundedSetPoints.put(rangeAction, (double) Math.round(rangeActionResult.getOptimizedSetPoint(rangeAction)));
+        }
```
Comment (pet-mit) [inline ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/linear_optimisation/IteratingLinearOptimizer.java:166]: improve readability, maybe by using two stream calls to build the new map (once on PstRangeActions, once on other RangeActions) ?

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java [hunk 5]
```diff
@@ -248,22 +252,47 @@ private void removeRangeActionsWithBiggestImpact(Set<RangeAction> rangeActions,
      * it will be considered greater.
      * If both RAs have the same priority, then absolute sensitivities will be compared.
      */
-    private int comparePrioritiesAndSensitivities(RangeAction ra1, RangeAction ra2, FlowCnec
-            cnec, SensitivityResult sensitivityResult) {
+    private int comparePrioritiesAndSensitivities(RangeAction ra1, RangeAction ra2, FlowCnec cnec, OptimizationResult optimizationResult) {
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java:255]: rename to comparePrioritiesAndPotentialGain

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java [hunk 6]
```diff
@@ -110,7 +114,7 @@ public void filterTsos() {
 
     Set<String> sortTsosToKeepBySensitivityAndGroupId(Set<String> activatedTsos, int maxTso) {
         List<RangeAction> rangeActionsSortedBySensitivity = rangeActionsToOptimize.stream()
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java:116]: rename to rangeActionsSortedByPotentialGain

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java [hunk 7]
```diff
@@ -248,22 +252,47 @@ private void removeRangeActionsWithBiggestImpact(Set<RangeAction> rangeActions,
...
+        double directMargin = cnec.getUpperBound(Side.LEFT, Unit.MEGAWATT).orElse(Double.POSITIVE_INFINITY) - optimizationResult.getFlow(cnec, Unit.MEGAWATT);
+        double oppositeMargin = optimizationResult.getFlow(cnec, Unit.MEGAWATT) - cnec.getLowerBound(Side.LEFT, Unit.MEGAWATT).orElse(Double.NEGATIVE_INFINITY);
+        double sensi = optimizationResult.getSensitivityValue(cnec, rangeAction, Unit.MEGAWATT);
+        double currentSetPoint = optimizationResult.getOptimizedSetPoint(rangeAction);
+        if (directMargin < oppositeMargin) {
+            if (sensi > 0) {
+                return sensi * (currentSetPoint - rangeAction.getMinAdmissibleSetpoint(prePerimeterSetPoints.get(rangeAction)));
+            } else {
+                return sensi * (currentSetPoint - rangeAction.getMaxAdmissibleSetpoint(prePerimeterSetPoints.get(rangeAction)));
+            }
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java:282]: can you add unit tests for this case?

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java (hunk 1 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java:231]: ok

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java (hunk 3 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:98]: I'm not a fan either. Sadly I didn't find a way with mocks... I believe we had the same problem with the tree parameters and the available range actions.

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java (hunk 5 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java:255]: ok

**Pull request merged**